python/pyspark/sql/functions.py

```diff
@@ -1847,8 +1848,7 @@ def __call__(self, *cols):
         return Column(jc)
 
 
-@since(1.3)
-def udf(f, returnType=StringType()):
+def _udf(f, returnType=StringType()):
     """Creates a :class:`Column` expression representing a user defined function (UDF).
 
     .. note:: The user-defined functions must be deterministic. Due to optimization,
```

Comment:
If we don't mind some discrepancy between docstring and usage (like I mentioned I believe that documenting internals with switching between decorator with partial and full application won't do us any good here) it can be pushed as an internal helper into `udf`.